refactor(core): add `debug` for i18n

mhevery opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
refactor(core): add human readable `debug` for i18n

I18n code breaks up internationalization into opCodes which are then stored
in arrays. To make it easier to debug the codebase this PR adds `debug`
property to the arrays which presents the data in human readable format.

AndrewKushnir commented on packages/core/src/render3/i18n_debug.ts:
```diff
@@ -0,0 +1,187 @@
+  const parser = new OpCodeParser(opcodes || (Array.isArray(this) ? this : []));
+  let lines: string[] = [];
+
+  function consumeOpCode(value: number): string {
```
Should this be a part of `OpCodeParser`?

petebacondarwin commented:
There are linting errors: https://app.circleci.com/pipelines/github/angular/angular/19025/workflows/4b3ada13-ed8a-4c8b-84a4-8a150e3c10d3/jobs/771505

Also this change appears to fail the code size check for `//integration:hello_world__closure_test`.

```
FAIL: Commit undefined uncompressed bundle exceeded expected size by 500 bytes or >1% (expected: 1209659, actual: 1211464).
```

petebacondarwin commented on packages/core/src/render3/i18n.ts:
```diff
@@ -1180,9 +1192,9 @@ function getCaseIndex(icuExpression: TIcu, bindingValue: string): number {
-  const updateCodes = [];
+  const createCodes: I18nMutateOpCodes[] = [];
+  const removeCodes: I18nMutateOpCodes[] = [];
+  const updateCodes: I18nUpdateOpCodes[] = [];
```
Don't these need to have the `attachDebugGetter()` called on them too?

mhevery replied:
Only `I18nUpdateOpCodes` need it (not `I18nUpdateOpCodes[]`)

petebacondarwin commented on packages/core/src/render3/i18n_debug.ts:
```diff
@@ -0,0 +1,187 @@
+/**
+ * Converts `I18nUpdateOpCodes` array into a human readable format.
+ *
+ * This function is attached to the `I18nUpdateOpCodes.debug` if `ngDevMode` is enabled. This
```
```suggestion
 * This function is attached to the `I18nUpdateOpCodes.debug` property if `ngDevMode` is enabled. This
```

petebacondarwin commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -6,18 +6,30 @@
  * `I18nMutateOpCode` defines OpCodes for `I18nMutateOpCodes` array.
  *
+ * OpCodes are efficient operations which can be applied to the DOM to update it. (For example to
+ * update to a new ICU case requires that we clean up previews elements and create new ones.)
```
```suggestion
 * update to a new ICU case requires that we clean up previous elements and create new ones.)
```

petebacondarwin commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -6,18 +6,30 @@
  *  2) Reference node index offset.
  *  3) The OpCode to execute.
  *
+ * pppp pppp pppp pppp rrrr rrrr rrrr riii
```
It might be worth clarifying in the list above the letter being used - particularly since `iii` doesn't seem to have a relationship to the term "OpCode".

E.g.

```
 *  1) Parent node index offset (`p`).
 *  2) Reference node index offset (`r`).
 *  3) The OpCode to execute (`i`).

petebacondarwin replied:
Moreover, it seems that the overall structure (`p...pr...riii`) is called an OpCode but also the action to execute (`iii`)  is also called an OpCode. It would be good to distinguish these two.

petebacondarwin commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -6,18 +6,30 @@
  *
+ * pppp pppp pppp pppp rrrr rrrr rrrr riii
+ * 3322 2222 2222 1111 1111 1110 0000 0000
+ * 1098 7654 3210 9876 5432 1098 7654 3210
```
It would be nice to show the decoded versions of these two opcode values.

mhevery replied:
done.

petebacondarwin commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -32,34 +44,60 @@ export const enum I18nMutateOpCode {
+}
+
+export function getRefFromI18nMutateOpCode(mergedCode: number): number {
+  if ((mergedCode & I18nMutateOpCode.MASK_REF) >>> I18nMutateOpCode.SHIFT_REF === 16384) debugger;
```
😱

petebacondarwin commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -32,34 +44,60 @@ export const enum I18nMutateOpCode {
+
 /**
- * Marks that the next string is for element.
+ * Marks that the next string is element name.
```
```suggestion
 * Marks that the next string is an element name.
```

petebacondarwin commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -256,7 +279,16 @@ export const enum I18nUpdateOpCode {
+   * on in production application.
+   */
+  debug?: string[]
+}
```
NIT: rather than repeating this for both `I18nUpdateOpCodes` and `I18nMutateOpCodes` how about something like:

```
export interface OpCodeDebugInfo {
  /**
   * Human readable representation of the opcodes.
   *
   * NOTE: This property only exists if `ngDevMode` is set to `true` and it is not present in
   * production. Its presence is purely to help debug issue in development, and should not be relied
   * on in production application.
   */
  debug?: string[]
}

export type I18nUpdateOpCodes = Array<string|number|SanitizerFn|null> & OpCodeDebugInfo;
export type I18nMutateOpCodes = Array<number|string|ELEMENT_MARKER|COMMENT_MARKER|null>  & OpCodeDebugInfo;
```

petebacondarwin commented on packages/core/test/render3/utils.ts:
```diff
@@ -0,0 +1,75 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/** Template string function that can be used to dedent a given string literal. */
```
```suggestion
/** Template string function that can be used to strip indentation from a given string literal. */
```

petebacondarwin commented on packages/core/test/render3/utils.ts:
```diff
@@ -0,0 +1,75 @@
+ * @param text
+ */
+function numOfWhiteSpaceLeadingChars(text: string): number {
+  return text.match(/^\s*/)![0].length;
```
Not a big issue since this is only for tests: This could get a bit muddled if some lines contain tab indents and others line only space indents, or even a random mix of the two.

mhevery replied:
Good, point, but hopefully the formatter will ensure that it is consistent.

petebacondarwin commented on packages/core/test/render3/utils.ts:
```diff
@@ -0,0 +1,75 @@
+  let actual: any = null;
+
+  matcher.asymmetricMatch = function(objectWithDebug: any) {
+    if (!objectWithDebug.debug) debugger;
```
Is this line to be kept? If so, it probably needs a tslint flag.

petebacondarwin reviewed (approved):
This looks great. Much nicer tests!

mhevery committed:
refactor(core): extract `icuSwitchCase`, `icuUpdateCase`, `removeNestedIcu`

Extract `icuSwitchCase`, `icuUpdateCase`, `removeNestedIcu` into
separate functions to align them with the `.debug` property text.

AndrewKushnir reviewed (commented):
Thanks for making debugging easier @mhevery 👍 I just left a few comments.

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +931,67 @@ function readUpdateOpCodes(
 }
 
+function icuUpdateCase(
+    tView: TView, tIcuIndex: number, nodeIndex: number, bindingsStartIndex: number,
```
done

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +931,67 @@ function readUpdateOpCodes(
+function icuUpdateCase(
+    tView: TView, tIcuIndex: number, nodeIndex: number, bindingsStartIndex: number,
+    icus: TIcu[]|null, lView: LView, caseCreated: boolean) {
+  let tIcu = icus![tIcuIndex];
```
done

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +931,67 @@ function readUpdateOpCodes(
+    icus: TIcu[]|null, lView: LView, caseCreated: boolean) {
+  let tIcu = icus![tIcuIndex];
+  let icuTNode = getTNode(tView, nodeIndex) as TIcuContainerNode;
+  if (icuTNode.activeCaseIndex !== null) {
```
done

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +931,67 @@ function readUpdateOpCodes(
+  icuTNode.activeCaseIndex = caseIndex !== -1 ? caseIndex : null;
+  if (caseIndex > -1) {
+    // Add the nodes for the new case
+    readCreateOpCodes(-1, tIcu.create[caseIndex], tView, lView);
```
done

atscott reviewed (approved):
reviewed-for: size-tracking

Based on the changes, I think the size increase should be debug only but it appears in the size tracking because of this TODO: "we should define ngDevMode to false in Closure, but --define only works in the global scope."

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -795,13 +803,15 @@ function readCreateOpCodes(
               assertDefined(
                   currentTNode!,
                   `You need to create or select a node before you can insert it into the DOM`);
+          debugger;
```
Leftover `debugger` statement.

mhevery replied:
removed

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -795,13 +803,15 @@ function readCreateOpCodes(
+          // Negative indices indicate that a given TNode is a sibling node, not a parent node
           // (see `i18nStartFirstPass` for additional information).
           const isParent = opCode >= 0;
+          // FIXME(COMMIT): This SHIFT_REF looks suspect as it does not have mask.
```
It looks like this comment should be updated?

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -890,13 +900,11 @@ function readUpdateOpCodes(
           if (opCode < 0) {
-            // It's a binding index whose value is negative
+            // Negative opCode represent `i18nExp` values offset.
+            debugger;
```
Leftover `debugger`.

mhevery replied:
removed

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +934,68 @@ function readUpdateOpCodes(
+function icuSwitchCase(
+    tView: TView, tIcuIndex: number, nodeIndex: number, icus: TIcu[]|null, lView: LView,
+    value: string): boolean {
+  let tIcu = icus![tIcuIndex];
```
I guess we can update the type of `icus` and remove `null` from there to avoid non-null assertion operator here (actually move it up where the `icuSwitchCase` function is called, similar to the `icuUpdateCase` function call above).

AndrewKushnir replied:
Another idea: it looks like `tIcuIndex` and `tIcu` arguments are passed to this function (and the `icuUpdateCase` one) to reference current `tIcu`, so we can just calculate it before calling this function and pass it as an argument (instead of passing the `tIcuIndex`).

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +934,68 @@ function readUpdateOpCodes(
+  if (activeIndex !== null) {
+    const nestedTIcu = icus![nodeIndex];
+    // FIXME(misko): the fact that we are adding items to parent list looks very suspect!
+    // addAllToArray(nestedTIcu.remove[activeIndex], removeCodes);
```
It looks like this code is needed, so we should un-comment it (there are some tests that are failing, probably because of that).

mhevery replied:
Uncommented for now. No tests fail if I leave it comment out. I will investigate later.

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -1018,6 +1045,7 @@ function removeNode(tView: TView, lView: LView, index: number, markAsDetached: b
  * @codeGenApi
  */
 export function ɵɵi18n(index: number, message: string, subTemplateIndex?: number): void {
+  debugger;
```
Leftover `debugger`.

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -1364,13 +1400,15 @@ function parseNodes(
           3,                                 // skip 3 opCodes if not changed
           -1 - nestedIcu.mainBinding,
           nestedIcuNodeIndex << I18nUpdateOpCode.SHIFT_REF | I18nUpdateOpCode.IcuSwitch,
+          // FIXME(misko): Index should be part of the opcode
```
Not sure if it's a WIP comment. If not, may be include a quick comment on why this should be moved?

mhevery replied:
It is part of things which I want to fix in follow up PRs.

AndrewKushnir commented on packages/core/src/render3/i18n_debug.ts:
```diff
@@ -0,0 +1,186 @@
+      case I18nUpdateOpCode.IcuUpdate:
+        return `icuUpdateCase(lView[${ref}] as Comment, ${parser.consumeNumber()})`;
+    }
+    throw 'unexpected OpCode';
```
```suggestion
    throw new Error('unexpected OpCode');
```
(to be consistent with the rest of the code)

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -30,36 +43,61 @@ export const enum I18nMutateOpCode {
+   */
+  // FIXME(misko): Why is this not used?
+  MASK_REF = 0b11111111111111000,
+  //           11111110000000000
+  //           65432109876543210
```
This should probably be removed?

mhevery replied:
Why? It will be needed in next PRs, and it shows how the bits are used.

AndrewKushnir commented on packages/core/src/render3/interfaces/i18n.ts:
```diff
@@ -30,36 +43,61 @@ export const enum I18nMutateOpCode {
   /**
    * Mask for OpCode
```
Should we update a comment here as well?

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +932,70 @@ function readUpdateOpCodes(
+function icuSwitchCase(
+    tView: TView, tIcuIndex: number, nodeIndex: number, icus: TIcu[], lView: LView,
+    value: string): boolean {
+  let tIcu = icus[tIcuIndex];
```
```suggestion
  const tIcu = icus[tIcuIndex];
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +932,70 @@ function readUpdateOpCodes(
+    tView: TView, tIcuIndex: number, nodeIndex: number, icus: TIcu[], lView: LView,
+    value: string): boolean {
+  let tIcu = icus[tIcuIndex];
+  let icuTNode = getTNode(tView, nodeIndex) as TIcuContainerNode;
```
```suggestion
  const icuTNode = getTNode(tView, nodeIndex) as TIcuContainerNode;
```

mhevery replied:
fixed

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +932,70 @@ function readUpdateOpCodes(
+    const removeCodes = tIcu.remove[icuTNode.activeCaseIndex];
+    for (let k = 0; k < removeCodes.length; k++) {
+      const removeOpCode = removeCodes[k] as number;
+      const nodeIndex = removeOpCode >>> I18nMutateOpCode.SHIFT_REF;
```
It looks like we shadow the `nodeIndex` argument here, so may be rename this const to `nodeOrIcuIndex` (as it can be both node and ICU index)?

mhevery replied:
renamed.

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +932,70 @@ function readUpdateOpCodes(
 
+function icuUpdateCase(
+    tView: TView, lView: LView, tIcuIndex: number, nodeIndex: number, bindingsStartIndex: number,
+    icus: TIcu[], caseCreated: boolean) {
```
I was thinking if we could rename the `icus` argument to `tIcus` (here and in other extracted functions). In this case the `tIcuIndex` argument name would be more self-explanatory.
```suggestion
    tIcus: TIcu[], caseCreated: boolean) {
```

mhevery replied:
done

AndrewKushnir reviewed (approved):
Thanks for updates @mhevery 👍

The change LGTM, just left a few small nits.

AndrewKushnir reviewed (approved):
Also adding LGTM for `size-tracking`...

Reviewed-for: size-tracking

mhevery commented:
[presubmit](http://test/OCL:324630738:BASE:324630765:1596475639276:d5a8448e)

AndrewKushnir commented on packages/core/src/render3/i18n.ts:
```diff
@@ -969,6 +932,70 @@ function readUpdateOpCodes(
+}
+
+function icuSwitchCase(
+    tView: TView, tIcuIndex: number, nodeIndex: number, icus: TIcu[], lView: LView,
```
@mhevery thanks for updating `icus` -> `tIcus` argument name in the `icuUpdateCase` fn. Could you plz also update it in the `icuSwitchCase` and `removeNestedIcu` fns?

mhevery replied:
Yes, of course. Done.

AndrewKushnir commented:
FYI, this PR didn't merge into the patch branch cleanly due to payload size limit file conflicts, I've create a patch-only version of this PR: https://github.com/angular/angular/pull/38370.
